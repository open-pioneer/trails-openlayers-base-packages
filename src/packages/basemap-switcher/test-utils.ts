// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { MapConfig, MapConfigProvider, MapRegistry } from "@open-pioneer/map";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import Stamen from "ol/source/Stamen";
import { createService } from "@open-pioneer/test-utils/services";
import { MapRegistryImpl } from "@open-pioneer/map/MapRegistryImpl";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { screen, waitFor } from "@testing-library/react";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}
export async function setupMap(options?: SimpleMapOptions) {
    const mapId = "test";
    const mapConfig: MapConfig = {
        initialView: {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        },
        projection: "EPSG:3857",
        layers: [
            {
                id: "b-1",
                title: "OSM",
                isBaseLayer: true,
                visible: false,
                layer: new TileLayer({
                    source: new OSM()
                })
            },
            {
                id: "b-2",
                title: "Toner",
                isBaseLayer: true,
                visible: true,
                layer: new TileLayer({
                    source: new Stamen({ layer: "toner" })
                })
            }
        ]
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

export function createPackageContextProviderProps(
    service: MapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "map.MapRegistry": service
        }
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
