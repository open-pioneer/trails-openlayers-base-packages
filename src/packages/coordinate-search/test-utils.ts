// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { act, fireEvent } from "@testing-library/react";

export function showDropdown(projSelect: HTMLElement) {
    // open dropdown to include options in snapshot; react-select creates list of options in dom after opening selection
    act(() => {
        fireEvent.keyDown(projSelect, { key: "ArrowDown" });
    });
}

export function getCurrentOptions() {
    return Array.from(
        document.getElementsByClassName("coordinate-input-select__option")
    ) as HTMLElement[];
}

export function getCurrentOptionValues(options: HTMLElement[]) {
    const values: (string | null)[] = [];
    for (const opt of options) {
        values.push(opt.textContent);
    }
    return values;
}

export function getClearButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const clearButton = buttonDiv.querySelector(".coordinate-input-clear-button");
    if (!clearButton) {
        throw new Error("clear button not rendered");
    }
    return clearButton;
}

export function getCopyButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const copyButton = buttonDiv.querySelector(".coordinate-input-copy-button");
    if (!copyButton) {
        throw new Error("copy button not rendered");
    }
    return copyButton;
}
