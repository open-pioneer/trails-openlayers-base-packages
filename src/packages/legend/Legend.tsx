// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { Box, Image, List, Text } from "@open-pioneer/chakra-integration";
import { Layer, MapModel, useMapModel, LayerBase, Sublayer } from "@open-pioneer/map";
import {
    ComponentType,
    FC,
    ReactNode,
    useCallback,
    useEffect,
    useRef,
    useState,
    useSyncExternalStore
} from "react";
import { CommonComponentProps, useCommonComponentProps } from "@open-pioneer/react-utils";
import { useIntl } from "open-pioneer:react-hooks";
import { WarningTwoIcon } from "@chakra-ui/icons";
import classNames from "classnames";

type LegendLayer = Layer | Sublayer;

/**
 * Properties of a legend item React component.
 */
export interface LegendItemComponentProps {
    /**
     * Related layer of the legend.
     */
    layer: LayerBase;
}

/**
 * Attributes of the legend attribute that can be specified on a layer.
 *
 * To show a legend for the layer, provide an imageUrl to an image to show
 * or provide a React component that will be rendered as a legend.
 */
export interface LegendItemAttributes {
    /**
     * (Optional) URL to an image that will be shown as a legend for the layer.
     */
    imageUrl?: string;

    /**
     * (Optional) React component that will be shown as customized legend for the layer.
     */
    Component?: ComponentType<LegendItemComponentProps>;
}

/**
 * These are special properties for the Legend.
 */
export interface LegendProps extends CommonComponentProps {
    /**
     * The id of the map.
     */
    mapId: string;

    /**
     * Specifies whether legend for active base layer is shown in the legend UI.
     * Defaults to `false`.
     */
    showBaseLayers?: boolean;
}

/**
 * The `Legend` component can be used to display the legend of layers that are visible in the map.
 */
export const Legend: FC<LegendProps> = (props) => {
    const { mapId, showBaseLayers = false } = props;
    const { containerProps } = useCommonComponentProps("legend", props);
    const { map } = useMapModel(mapId);

    return (
        <Box {...containerProps}>
            {map ? <LegendList map={map} showBaseLayers={showBaseLayers} /> : null}
        </Box>
    );
};

function LegendList(props: { map: MapModel; showBaseLayers: boolean }): JSX.Element {
    const { map, showBaseLayers } = props;

    const layers = useLayers(map);
    const legendListItems: ReactNode[] = layers.map((layer) => {
        return (
            <LegendItem key={layer.id} layer={layer} showBaseLayers={showBaseLayers}></LegendItem>
        );
    });

    return (
        <List
            // Note: not using UnorderedList because it adds default margins
            as="ul"
            className="legend-layer-list"
            listStyleType="none"
            spacing={2}
        >
            {legendListItems}
        </List>
    );
}

function LegendItem(props: { layer: LegendLayer; showBaseLayers: boolean }): ReactNode {
    const { layer, showBaseLayers } = props;
    const { isVisible } = useVisibility(layer);
    const sublayers = useSublayers(layer);

    if (!isVisible) {
        return undefined;
    }

    // '!("parentLayer" in layer)' checks if the layer is no sublayer
    if (!showBaseLayers && !("parentLayer" in layer) && layer.isBaseLayer) {
        return undefined;
    }

    // legend items for all sublayers
    const childItems: ReactNode[] = [];
    if (sublayers?.length) {
        sublayers.forEach((sublayer) => {
            childItems.push(
                <LegendItem key={sublayer.id} layer={sublayer} showBaseLayers={showBaseLayers} />
            );
        });
    }

    return (
        <>
            <LegendContent layer={layer} showBaseLayers={showBaseLayers} />
            {childItems}
        </>
    );
}

function LegendContent(props: { layer: LegendLayer; showBaseLayers: boolean }) {
    const intl = useIntl();

    const { layer, showBaseLayers } = props;
    const legendAttributes = useLegendAttributes(layer);
    const legendUrl = useLegend(layer);
    let renderedComponent: ReactNode | undefined;

    if (legendAttributes?.Component) {
        renderedComponent = <legendAttributes.Component layer={layer} />;
    } else if (legendAttributes?.imageUrl) {
        renderedComponent = <LegendImage layer={layer} imageUrl={legendAttributes.imageUrl} />;
    } else {
        if (legendUrl) {
            renderedComponent = <LegendImage layer={layer} imageUrl={legendUrl} />;
        }
    }

    const isBaseLayer = !("parentLayer" in layer) && layer.isBaseLayer;

    return renderedComponent ? (
        <Box as="li" className={classNames("legend-item", `layer-${slug(layer.id)}`)}>
            {showBaseLayers && isBaseLayer ? (
                /* Render additional text, if layer is a configured basemap */
                <Text as="b">{intl.formatMessage({ id: "basemapLabel" })}</Text>
            ) : null}
            {renderedComponent}
        </Box>
    ) : undefined;
}

function LegendImage(props: { imageUrl: string; layer: LegendLayer }) {
    const intl = useIntl();

    const { layer, imageUrl } = props;

    return (
        <Box>
            <Text>{layer.title}</Text>
            <Image
                maxW="none"
                maxH="none"
                src={imageUrl}
                alt={intl.formatMessage({ id: "altLabel" }, { layerName: layer.title })}
                className={"legend-item__image"}
                fallbackStrategy={"onError"}
                fallback={
                    <Box>
                        <Text>
                            <WarningTwoIcon me={2} />
                            {intl.formatMessage({ id: "fallbackLabel" })}
                        </Text>
                    </Box>
                }
            />
        </Box>
    );
}

function useLegend(layer: LayerBase): string | undefined {
    const getSnapshot = useCallback(() => layer.legend, [layer]);
    const subscribe = useCallback(
        (cb: () => void) => {
            const resource = layer.on("changed:legend", cb);
            return () => resource.destroy();
        },
        [layer]
    );

    return useSyncExternalStore(subscribe, getSnapshot);
}

/** Returns the top level operation layers (without LayerGroups). */
function useLayers(map: MapModel): Layer[] {
    const subscribe = useCallback(
        (cb: () => void) => {
            const resource = map.layers.on("changed", cb);
            return () => resource.destroy();
        },
        [map]
    );
    const getValue = useCallback(() => {
        let layers = map.layers.getAllLayers({ sortByDisplayOrder: true }) ?? [];
        layers = layers.reverse();
        return layers;
    }, [map]);
    return useCachedExternalStore(subscribe, getValue);
}

/** Returns the sublayers of the given layer (or undefined, if the sublayer cannot have any). */
function useSublayers(layer: LayerBase): Sublayer[] | undefined {
    const subscribe = useCallback(
        (cb: () => void) => {
            const resource = layer.sublayers?.on("changed", cb);
            return () => resource?.destroy();
        },
        [layer]
    );
    const getValue = useCallback((): Sublayer[] | undefined => {
        const sublayers = layer.sublayers;
        if (!sublayers) {
            return undefined;
        }

        let layers = layer.sublayers?.getSublayers({ sortByDisplayOrder: true });
        layers = layers.reverse();
        return layers;
    }, [layer]);
    return useCachedExternalStore(subscribe, getValue);
}

/**
 * This hooks wraps an external store that does not cache its own values, i.e.
 * it may return a different value each time from `getValue()`.
 *
 * The results returned from `getValue()` are cached locally; the cache
 * is only invalidated on re-subscription or if a change event has been observed.
 */
function useCachedExternalStore<T>(
    subscribe: (onStoreChanged: () => void) => () => void,
    getValue: () => T
): T {
    const cachedValue = useRef<{ value: T } | undefined>();

    const cachedSubscribe = useCallback(
        (cb: () => void) => {
            const cleanup = subscribe(() => {
                // Reset cache on change
                cachedValue.current = undefined;
                cb();
            });
            return () => {
                // Reset cache when (re-) subscribing
                cachedValue.current = undefined;
                cleanup();
            };
        },
        [subscribe]
    );
    const cachedGetSnapshot = useCallback(() => {
        // Return cached values if still up to date (see resets above).
        if (cachedValue.current) {
            return cachedValue.current.value;
        }

        // Compute values and cache the result.
        const value = getValue();
        cachedValue.current = { value };
        return value;
    }, [getValue]);
    return useSyncExternalStore(cachedSubscribe, cachedGetSnapshot);
}

/** Returns the layer's current visibility. */
function useVisibility(layer: LayerBase): {
    isVisible: boolean;
} {
    const getSnapshot = useCallback(() => layer.visible, [layer]);
    const subscribe = useCallback(
        (cb: () => void) => {
            const resource = layer.on("changed:visible", cb);
            return () => resource.destroy();
        },
        [layer]
    );
    const isVisible = useSyncExternalStore(subscribe, getSnapshot);

    return {
        isVisible
    };
}

function useLegendAttributes(layer: LayerBase) {
    const [legendAttributes, setLegendAttributes] = useState<LegendItemAttributes | undefined>(
        undefined
    );

    useEffect(() => {
        setLegendAttributes(layer.attributes.legend as LegendItemAttributes | undefined);

        const resource = layer.on("changed:attributes", () => {
            setLegendAttributes(layer.attributes.legend as LegendItemAttributes | undefined);
        });
        return () => resource.destroy();
    }, [layer]);

    return legendAttributes;
}

function slug(id: string) {
    return id
        .toLowerCase()
        .replace(/[^a-z0-9 -]/g, "")
        .replace(/\s+/g, "-")
        .replace(/-+/g, "-");
}
