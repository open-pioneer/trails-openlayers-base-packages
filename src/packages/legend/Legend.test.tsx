// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { expect, it, vi } from "vitest";
import { createServiceOptions, setupMap } from "@open-pioneer/map-test-utils";
import TileLayer from "ol/layer/Tile";
import { act, render, screen, waitFor } from "@testing-library/react";
import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import { Legend, LegendItemComponentProps } from "./Legend";
import { Box, Image, Text } from "@open-pioneer/chakra-integration";
import { SimpleLayer, WMSLayer } from "@open-pioneer/map";
import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const THIS_DIR = dirname(fileURLToPath(import.meta.url));
const WMTS_CAPAS = readFileSync(resolve(THIS_DIR, "./test-data/SimpleWMSCapas.xml"), "utf-8");

// Happy dom does not have an XML parser
import jsdom from "jsdom";
window.DOMParser = new jsdom.JSDOM().window.DOMParser;

const LEGEND_ITEM_CLASS = ".legend-item";
const LEGEND_IMAGE_CLASS = ".legend-item__image";

it("should successfully create a legend component", async () => {
    const { mapId, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            },
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({})
            },
            {
                title: "Layer 2",
                id: "layer-2",
                olLayer: new TileLayer({})
            }
        ]
    });
    await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Legend mapId={mapId} data-testid="legend" />
        </PackageContextProvider>
    );

    const legendDiv = await findLegend();
    expect(legendDiv).toMatchSnapshot();
});

it("should successfully show legend for imageUrl configuration", async () => {
    const { mapId, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            },
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        imageUrl: "https://avatars.githubusercontent.com/u/121286957?s=200&v=4"
                    }
                }
            },
            {
                title: "Layer 2",
                id: "layer-2",
                olLayer: new TileLayer({})
            }
        ]
    });
    await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Legend mapId={mapId} data-testid="legend" />
        </PackageContextProvider>
    );

    const legendDiv = await findLegend();
    await waitForLegendItem(legendDiv);
    // do snapshot test to test if layer title and image are shown correctly
    expect(legendDiv).toMatchSnapshot();
});

it("should successfully show legend for Component configuration", async () => {
    const dotStyle = {
        height: "25px",
        width: "25px",
        borderColor: "#4cb3ff",
        borderWidth: "3px",
        borderRadius: "50%",
        display: "inline-block "
    };

    const { mapId, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            },
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        Component: function CustomLegend(props: LegendItemComponentProps) {
                            return (
                                <Box>
                                    <Text>{props.layer.title}</Text>
                                    <Box style={dotStyle}></Box>
                                </Box>
                            );
                        }
                    }
                }
            },
            {
                title: "Layer 2",
                id: "layer-2",
                olLayer: new TileLayer({})
            }
        ]
    });
    await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Legend mapId={mapId} data-testid="legend" />
        </PackageContextProvider>
    );

    const legendDiv = await findLegend();
    await waitForLegendItem(legendDiv);
    expect(legendDiv).toMatchSnapshot();
});

it("does not show a legend for basemaps by default", async () => {
    const { mapId, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true,
                attributes: {
                    "legend": {
                        imageUrl: "https://basemap-url.com/"
                    }
                }
            },
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        imageUrl: "https://avatars.githubusercontent.com/u/121286957?s=200&v=4"
                    }
                }
            },
            {
                title: "Layer 2",
                id: "layer-2",
                olLayer: new TileLayer({})
            }
        ]
    });
    await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Legend mapId={mapId} data-testid="legend" />
        </PackageContextProvider>
    );

    const legendDiv = await findLegend();
    await waitForLegendItem(legendDiv);

    const images = await getLegendImages(legendDiv);
    expect(images.length).toBe(1);

    const src = images[0]?.getAttribute("src");
    expect(src).not.toBe("https://basemap-url.com/");
});

it("shows a legend for active basemap if showBaseLayers is configured to be true", async () => {
    const { mapId, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true,
                attributes: {
                    "legend": {
                        imageUrl: "https://basemap-url.com/"
                    }
                }
            },
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        imageUrl: "https://avatars.githubusercontent.com/u/121286957?s=200&v=4"
                    }
                }
            },
            {
                title: "Layer 2",
                id: "layer-2",
                olLayer: new TileLayer({})
            }
        ]
    });
    await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Legend mapId={mapId} data-testid="legend" showBaseLayers={true} />
        </PackageContextProvider>
    );

    const legendDiv = await findLegend();
    await waitForLegendItem(legendDiv);

    const images = await getLegendImages(legendDiv, 2);
    expect(images.length).toBe(2);

    const src = images[1]?.getAttribute("src");
    expect(src).toBe("https://basemap-url.com/");
});

it("shows correct legend entries for nested WMSSublayers", async () => {
    const { mapId, registry } = await setupMap({
        layers: [createLayerWithNestedSublayers()],
        fetch: vi.fn(async () => {
            return new Response(WMTS_CAPAS, {
                status: 200
            });
        })
    });
    await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Legend mapId={mapId} data-testid="legend" />
        </PackageContextProvider>
    );

    const legendDiv = await findLegend();
    await waitForLegendItem(legendDiv);
    const images = await getLegendImages(legendDiv, 4);

    // Ensure that no legend is created for sublayer without `name` prop and for sublayer with empty legend component
    expect(images[0]?.getAttribute("src")).toBe("https://fake.legend.url/sublayer4_1.png");
    expect(images[1]?.getAttribute("src")).toBe("https://fake.legend.url/sublayer4_2.png");
    expect(images[2]?.getAttribute("src")).toBe("https://fake.legend.url/sublayer3_2.png");
    // legend src is extract from the capabilities
    expect(images[3]?.getAttribute("src")).toBe("http://www.university.edu/legends/atlas.gif");
});

it("shows legend entries in correct order", async () => {
    const { mapId, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            },
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        imageUrl: "https://fake.image.url/layer-1.png"
                    }
                }
            },
            {
                title: "Layer 2",
                id: "layer-2",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        imageUrl: "https://fake.image.url/layer-2.png"
                    }
                }
            },
            {
                title: "Layer 3",
                id: "layer-3",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        Component: function CustomLegend(props: LegendItemComponentProps) {
                            return (
                                <Box>
                                    <Text>{props.layer.title}</Text>
                                    <Box>
                                        <Image
                                            className="legend-item__image"
                                            src="https://fake.image.url/layer-3.png"
                                        ></Image>
                                    </Box>
                                </Box>
                            );
                        }
                    }
                }
            }
        ]
    });
    await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Legend mapId={mapId} data-testid="legend" />
        </PackageContextProvider>
    );

    const legendDiv = await findLegend();
    await waitForLegendItem(legendDiv);

    const images = await getLegendImages(legendDiv, 3);
    expect(images.length).toBe(3);
    expect(images[0]?.getAttribute("src")).toBe("https://fake.image.url/layer-3.png");
    expect(images[1]?.getAttribute("src")).toBe("https://fake.image.url/layer-2.png");
    expect(images[2]?.getAttribute("src")).toBe("https://fake.image.url/layer-1.png");
});

it("shows legend entries only for visible layers", async () => {
    const { mapId, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            },
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        imageUrl: "https://avatars.githubusercontent.com/u/121286957?s=200&v=4"
                    }
                }
            },
            {
                title: "Layer 2",
                id: "layer-2",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        imageUrl: "https://not-visbile-layer.com/"
                    }
                },
                visible: false
            }
        ]
    });
    await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Legend mapId={mapId} data-testid="legend" />
        </PackageContextProvider>
    );

    const legendDiv = await findLegend();
    await waitForLegendItem(legendDiv);

    const images = await getLegendImages(legendDiv);
    expect(images.length).toBe(1);

    const src = images[0]?.getAttribute("src");
    expect(src).not.toBe("https://not-visbile-layer.com/");
});

it("includes the layer id in the legend item's class list", async () => {
    const { mapId, registry } = await setupMap({
        layers: [
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        imageUrl: "https://avatars.githubusercontent.com/u/121286957?s=200&v=4"
                    }
                }
            }
        ]
    });
    await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Legend mapId={mapId} data-testid="legend" />
        </PackageContextProvider>
    );

    const legendDiv = await findLegend();
    const firstLegendItem = await waitForLegendItem(legendDiv);
    expect(firstLegendItem.classList.contains("layer-layer-1")).toBe(true);
});

it("reacts to changes in layer visibility", async () => {
    const { mapId, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            },
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        imageUrl: "https://fake.image.url/layer-1.png"
                    }
                }
            },
            {
                title: "Layer 2",
                id: "layer-2",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        Component: function CustomLegend(props: LegendItemComponentProps) {
                            return (
                                <Box>
                                    <Text>{props.layer.title}</Text>
                                    <Box>
                                        <Image
                                            className="legend-item__image"
                                            src="https://fake.image.url/layer-2.png"
                                        ></Image>
                                    </Box>
                                </Box>
                            );
                        }
                    }
                }
            }
        ]
    });
    const map = await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Legend mapId={mapId} data-testid="legend" />
        </PackageContextProvider>
    );

    // First check
    const legendDiv = await findLegend();
    await waitForLegendItem(legendDiv);

    const images = await getLegendImages(legendDiv, 2);
    expect(images.length).toBe(2);

    // Set visible to false
    const layers = map.layers.getOperationalLayers();
    act(() => {
        layers[0]?.setVisible(false);
    });

    // Second check
    const nextLegendDiv = await findLegend();
    await waitForLegendItem(legendDiv);

    const nextImages = await getLegendImages(nextLegendDiv);
    expect(nextImages.length).toBe(1);
});

it("reacts to changes in layer legend attributes", async () => {
    const { mapId, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            },
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        imageUrl: "https://fake.image.url/layer-1.png"
                    }
                }
            }
        ]
    });
    const map = await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Legend mapId={mapId} data-testid="legend" />
        </PackageContextProvider>
    );

    // First check
    const legendDiv = await findLegend();
    await waitForLegendItem(legendDiv);

    const images = await getLegendImages(legendDiv);
    expect(images[0]?.getAttribute("src")).toBe("https://fake.image.url/layer-1.png");

    // Update attributes
    const layers = map.layers.getOperationalLayers();
    act(() => {
        layers[0]?.updateAttributes({
            legend: { imageUrl: "https://fake.image.url/new_layer.png" }
        });
    });

    // Second check
    const nextLegendDiv = await findLegend();
    await waitForLegendItem(legendDiv);

    const nextImages = await getLegendImages(nextLegendDiv);
    expect(nextImages[0]?.getAttribute("src")).toBe("https://fake.image.url/new_layer.png");
});

it("reacts to changes in the layer composition", async () => {
    const { mapId, registry } = await setupMap({
        layers: [
            {
                title: "Base layer",
                id: "base-layer",
                olLayer: new TileLayer({}),
                isBaseLayer: true
            },
            {
                title: "Layer 1",
                id: "layer-1",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        imageUrl: "https://fake.image.url/layer-1.png"
                    }
                }
            }
        ]
    });
    const map = await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });

    render(
        <PackageContextProvider services={injectedServices}>
            <Legend mapId={mapId} data-testid="legend" />
        </PackageContextProvider>
    );

    const layers = map.layers.getOperationalLayers();
    expect(layers.length).toBe(1);

    act(() => {
        map.layers.addLayer(
            new SimpleLayer({
                title: "Layer 2",
                id: "layer-2",
                olLayer: new TileLayer({}),
                attributes: {
                    "legend": {
                        imageUrl: "https://fake.image.url/layer-2.png"
                    }
                }
            })
        );
    });

    const nextLayers = map.layers.getOperationalLayers();
    expect(nextLayers.length).toBe(2);

    const legendDiv = await findLegend();
    await waitForLegendItem(legendDiv);

    const images = await getLegendImages(legendDiv, 2);
    expect(images[0]?.getAttribute("src")).toBe("https://fake.image.url/layer-2.png");
    expect(images[1]?.getAttribute("src")).toBe("https://fake.image.url/layer-1.png");
});

async function findLegend() {
    const legendDiv = await screen.findByTestId("legend");
    return legendDiv;
}

async function waitForLegendItem(legendDiv: HTMLElement) {
    return await waitFor(() => {
        const legendItem = legendDiv.querySelector(LEGEND_ITEM_CLASS);
        if (!legendItem) {
            throw new Error("legend item not mounted");
        }

        return legendItem;
    });
}

async function getLegendImages(legendDiv: HTMLElement, expectedCount = 1) {
    return await waitFor(() => {
        const legendImages = legendDiv.querySelectorAll(LEGEND_IMAGE_CLASS);
        if (legendImages.length < expectedCount) {
            throw new Error(
                `expected at least ${expectedCount} legend image(s), got only ${legendImages.length}`
            );
        }

        return Array.from(legendImages);
    });
}

function createLayerWithNestedSublayers() {
    return new WMSLayer({
        title: "Nested Layer",
        visible: true,
        url: "https://fake.wms.url/service",
        sublayers: [
            {
                title: "Sublayer 1",
                sublayers: [
                    {
                        title: "Sublayer 2",
                        sublayers: [
                            {
                                title: "Sublayer 3.2",
                                // legend for nested layer group
                                attributes: {
                                    "legend": {
                                        imageUrl: "https://fake.legend.url/sublayer3_2.png"
                                    }
                                },
                                sublayers: [
                                    {
                                        name: "sublayer4_4",
                                        title: "Sublayer 4.4",
                                        attributes: {
                                            "legend": {
                                                Component: function EmptyLegend() {
                                                    return null;
                                                }
                                            }
                                        }
                                    },
                                    {
                                        name: "sublayer4_3",
                                        title: "Sublayer 4.3"
                                    }
                                ]
                            },
                            {
                                title: "Sublayer 3.1",
                                sublayers: [
                                    {
                                        name: "sublayer4_2",
                                        title: "Sublayer 4.2",
                                        attributes: {
                                            "legend": {
                                                imageUrl: "https://fake.legend.url/sublayer4_2.png"
                                            }
                                        }
                                    },
                                    {
                                        name: "sublayer4_1",
                                        title: "Sublayer 4.1",
                                        attributes: {
                                            "legend": {
                                                imageUrl: "https://fake.legend.url/sublayer4_1.png"
                                            }
                                        }
                                    }
                                ]
                            }
                        ]
                    }
                ]
            }
        ]
    });
}
