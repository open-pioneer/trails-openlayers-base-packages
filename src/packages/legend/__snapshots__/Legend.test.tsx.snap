// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should successfully create a legend component 1`] = `
<div
  class="legend css-0"
  data-testid="legend"
  data-theme="light"
>
  <ul
    class="legend-layer-list css-bhb3xr"
    data-theme="light"
    role="list"
  />
</div>
`;

exports[`should successfully show legend for Component configuration 1`] = `
<div
  class="legend css-0"
  data-testid="legend"
  data-theme="light"
>
  <ul
    class="legend-layer-list css-bhb3xr"
    data-theme="light"
    role="list"
  >
    <li
      class="legend-item layer-layer-1 css-0"
      data-theme="light"
    >
      <div
        class="css-0"
        data-theme="light"
      >
        <p
          class="chakra-text css-0"
          data-theme="light"
        >
          Layer 1
        </p>
        <div
          class="css-0"
          data-theme="light"
          style="height: 25px; width: 25px; border-color: #4cb3ff; border-width: 3px; border-radius: 50%; display: inline-block;"
        />
      </div>
    </li>
  </ul>
</div>
`;

exports[`should successfully show legend for imageUrl configuration 1`] = `
<div
  class="legend css-0"
  data-testid="legend"
  data-theme="light"
>
  <ul
    class="legend-layer-list css-bhb3xr"
    data-theme="light"
    role="list"
  >
    <li
      class="legend-item layer-layer-1 css-0"
      data-theme="light"
    >
      <div
        class="css-0"
        data-theme="light"
      >
        <p
          class="chakra-text css-0"
          data-theme="light"
        >
          Layer 1
        </p>
        <img
          alt="altLabel"
          class="legend-item__image css-12ffw4m"
          data-theme="light"
          src="https://avatars.githubusercontent.com/u/121286957?s=200&v=4"
        />
      </div>
    </li>
  </ul>
</div>
`;
